doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,435 @@
+| 0 | Valid in 0-RTT packets|
+| 1 | Valid in 1-RTT packets|
+
+##### Spec
+
+_Spec_ is defined as:
+
+| Spec | Description |
+| :---: | --- |
+| N | Not ack-eliciting. |
+| C | does not count toward bytes in flight for congestion control purposes. |
+| P | Can be used to probe new network paths during connection migration. |
+| F | The contents of frames with this marking are flow controlled. |
+
+For `C`, `N` and `P`, the entire packet must consist of only frames with the
+marking for the packet to qualify for it.  For example, a packet with an ACK
+frame and a _stream_ frame would qualify for neither the `C` or `N` markings.
+
+### Packets
+
+Frames are coalesced into packets which are then sent by the record layer.
+The `packet_header` is a pointer to the leading bytes of the packet.
+The `frames` are pointers to the individual frames that make up the
+packet's body.
+It is expected that the record layer will encrypt from the `packet_header` and
+`frames` directly without a copy.
+
+```c
+enum packet_validity_e {
+    QUIC_PACKET_INITIAL,
+    QUIC_PACKET_HANDSHAKE,
+    QUIC_PACKET_0_RTT,
+    QUIC_PACKET_1_RTT
+};
+
+typedef enum packet_validity_e PACKET_VALIDITY;
+
+struct ossl_quic_packet_st {
+    QUIC_CONNECTION *conn;
+    unsigned char *packet_header;
+    size_t packet_header_length;
+    STACK_OF(OSSL_QUIC_TXP_FRAME) *frames;
+
+    QUIC_PN packet_number; /* RFC 9000 12.3 */
+    size_t packet_length;
+
+    /* Pkts - might be better as one field as per ACK Manager */
+    unsigned int initial : 1;
+    unsigned int handshake : 1;
+    unsigned int rtt0 : 1;
+    unsigned int rtt1 : 1;
+    /* Pkts options */
+    PACKET_VALIDITY validity;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+#### Notes
+
+- `QUIC_PN` defined as uint64_t in ACK Manager.
+- The `QUIC_ACKM_TX_PKT` structure from the ACK Manager overlaps this
+  significantly and the two could be merged.
+- Do we need the distinction between 0-rtt and 1-rtt when both are in
+  the Application Data number space?
+- 0-RTT packets can morph into 1-RTT packets and this needs to be handled by
+  the packetiser.
+- the `packet_header` will become a structure / union most likely.
+
+Interactions
+------------
+
+The packetiser needs to interact with other modules.  This defines the APIs
+by which it does so.
+
+Frames are passed to the packetiser on a per stream basis.
+The frames must be fully formed.  By passing a frame to this function,
+ownership is passed to the packetiser which queues the frames for later
+sending by the record layer.
+
+```c
+int ossl_quic_packetiser_buffer_frame(OSSL_QUIC_TX_PACKETISER *tx,
+                                      QUIC_STREAM *stream,
+                                      const OSSL_QUIC_FRAME *frame,
+                                      size_t frame_length);
```

Comment:
Is `QUIC_STREAM` the correct object to pass here? Isn't stream "pulled" in via the `ossl_quic_get_app_data` function described below? I assume this function is for other types of frame. Wouldn't `QUIC_CONNECTION` be more appropriate?

Reply:
This comment is outstanding

Reply:
Are frames written to a stream or a connection?

Reply:
Application data is written to streams. Frames are written to a connection IMO.